src/core/experimental/rcptr.d

```diff
@@ -0,0 +1,208 @@
+*/
+module core.experimental.rcptr;
+
+///
+unittest
+{
+    import core.stdc.stdlib : calloc;
+    auto allocInts = (size_t count) => cast(int*) calloc(count, int.sizeof);
+
+    __rcptr!int a; //default ctor
+    assert(a.get is null);
+
+    auto b = __rcptr!int(allocInts(10)); //ptr ctor
+
+    {
+        auto c = a; //copy ctor
+        assert(c.get == a.get);
+        assert(c.get != b.get);
+
+        c = b; //opAssign
+        assert(c.get != a.get);
+        assert(c.get == b.get);
+    }
+
+    assert(a.count is null);
+    assert(*b.count == 0);
+}
+
+/**
+A reference counted shared pointer that can be used to implement reference
+counted data structures.
+
+The reference count is automatically incremented/decremented on assignment,
+copy (construction), and destruction. When there are no more references to the
+pointer, the reference count is automatically deallocated and the referenced
+pointer is `free`d.
+
+Implementation: The internal implementation of `__rcptr` uses `malloc`/`free`.
+*/
+struct __rcptr(T)
+{
+    alias CounterType = int;
+
+    private T* ptr = null;
+    private shared(CounterType)* count = null;
+
+    /**
+    Creates a new `__rcptr` instance, tracking the provided pointer.
+
+    This implies that the ownership of the pointer is transferred to
+    `__rcptr`.
+
+    Params:
+         ptr = pointer to memory to be managed by `__rcptr`
+    */
+    this(inout(T*) ptr) inout
+    {
+        this.ptr = ptr;
+
+        if (ptr !is null)
+        {
+            import core.memory : pureCalloc;
+
+            // We use `calloc` so we don't have to manually initialise count/addRef
+            count = (() @trusted inout => cast(typeof(count)) pureCalloc(1, CounterType.sizeof))();
+        }
+    }
+
+    /**
+    Get the owned pointer.
+
+    Returns:
+        The pointer owned by this instance of `__rcptr`.
+    */
+    @system
+    inout(T*) get() inout
+    {
+        return ptr;
+    }
+
+    ~this()
+    {
+        delRef();
+    }
+
+    ///
+    void opAssign(__rcptr!T rhs)
+    {
+        if (rhs.count == count)
+        {
+            return;
+        }
+
+        delRef();
+
+        ptr = rhs.ptr;
+        count = rhs.count;
+
+        addRef();
+    }
+
+    ///
+    this(scope ref inout __rcptr!T rhs) inout
+    {
+        ptr = rhs.ptr;
+        count = rhs.count;
+
+        addRef();
+    }
+
+    private void addRef() inout
+    {
+        import core.atomic : atomicOp;
+
+        if (ptr is null)
+        {
+            return;
+        }
+
+        () @trusted { atomicOp!"+="(*(cast(shared(CounterType)*) count), 1); } ();
+    }
+
+    private void delRef()
+    {
+        import core.atomic : atomicOp;
+
+        if (ptr is null)
+        {
+            return;
+        }
+
+        // The counter is left at -1 when this was the last reference
+        // (i.e. the counter is 0-based, because we use calloc)
+        if (atomicOp!"=="(*count, 0) || atomicOp!"-="(*count, 1) == -1)
```

Comment:
I don't know what `atomicOp!"=="` actually is, but I think this is not right either.
It should be a non-atomic comparison:
```d
if (atomicLoad!(MemoryOrder.raw)(*count) == 0 || atomicOp!"-="(*count, 1) == -1)
```
The decrement should be relaxed too.

Reply:
Looks like it should be right like this? https://github.com/dlang/druntime/blob/3c9b39f3d9f6575d8df5289747f7e6d0f0680b6b/src/core/atomic.d#L819

Reply:
I'm not sure what or why you just linked to that? Or what you mean by 'like this'?

Reply:
Oh , maybe I see what you're saying...
Considering this code:
```
         TailShared!T get = atomicLoad!(MemoryOrder.raw)( val ); 
         mixin( "return get " ~ op ~ " mod;" ); 
```
It's clear that the comparison is not atomic; despite the name of the function...

So yeah you're right, but while that may be true, I would recommend NOT calling a function called `atomicOp` that performs a non-atomic op... it's misleading at best, and for my money, ugly and harder to understand.

Reply:
> So yeah you're right, but while that may be true, I would recommend NOT calling a function called `atomicOp` that performs a non-atomic op... it's misleading at best, and for my money, ugly and harder to understand.

That's fair, especially considering `atomicOp`'s documentation doesn't specify it.

Reply:
It's really not a great library.